utility_analysis_new/utility_analysis.py

```diff
@@ -26,19 +31,62 @@ class UtilityAnalysisOptions:
 
 
 def perform_utility_analysis(col,
+                             backend: pipeline_backend.PipelineBackend,
                              options: UtilityAnalysisOptions,
                              data_extractors: pipeline_dp.DataExtractors,
                              public_partitions=None):
     """Performs utility analysis for DP aggregations.
 
   Args:
     col: collection where all elements are of the same type.
+    backend: PipelineBackend with which the utility analysis will be run.
     options: options for utility analysis.
     data_extractors: functions that extract needed pieces of information
           from elements of 'col'.
     public_partitions: A collection of partition keys that will be present
           in the result. If not provided, the utility analysis with private
           partition selection will be performed.
   """
-    raise NotImplementedError(
-        "perform_utility_analysis is not implemented yet.")
+    budget_accountant = pipeline_dp.NaiveBudgetAccountant(
+        total_epsilon=options.eps, total_delta=options.delta)
+    engine = dp_engine.UtilityAnalysisEngine(
+        budget_accountant=budget_accountant, backend=backend)
+    result = engine.aggregate(col,
+                              params=options.aggregate_params,
+                              data_extractors=data_extractors,
+                              public_partitions=public_partitions)
+    budget_accountant.compute_budgets()
+
+    def create_aggregate_error_compound_combiner(
+            aggregate_params: pipeline_dp.AggregateParams,
+            error_quantiles: List[float],
+            public_partitions: bool) -> combiners.CompoundCombiner:
+        internal_combiners = []
+        if not public_partitions:
+            internal_combiners.append(
+                utility_analysis_combiners.
+                PrivatePartitionSelectionAggregateErrorMetricsCombiner(
+                    None, error_quantiles))
+        if pipeline_dp.Metrics.COUNT in aggregate_params.metrics:
+            internal_combiners.append(
+                utility_analysis_combiners.CountAggregateErrorMetricsCombiner(
+                    None, error_quantiles))
+        return utility_analysis_combiners.AggregateErrorMetricsCompoundCombiner(
+            internal_combiners, return_named_tuple=False)
+
+    aggregate_error_combiners = create_aggregate_error_compound_combiner(
+        options.aggregate_params, [0.1, 0.5, 0.9, 0.99],
+        public_partitions != None)
+    # TODO: Implement combine_accumulators (w/o per_key)
+    keyed_by_same_key = backend.map(result, lambda v: (None, v[1]),
+                                    "Rekey partitions by the same key")
+    accumulators = backend.map_values(
```

Comment:
Nit: add comments with types of collection elements after each of the transformation, it's very helpful to know them 
[Example](https://github.com/OpenMined/PipelineDP/blob/4540083dfbab6377b28c3912dfc5ef70a8141075/pipeline_dp/dp_engine.py#L127)